pkg/controller/gcp/storage/bucket.go

```diff
@@ -0,0 +1,315 @@
+	return &bucketSyncDeleter{
+		createupdater: newBucketCreateUpdater(sc, cc, b, pID),
+		Client:        sc,
+		client:        cc,
+		object:        b,
+	}
+}
+
+func (bh *bucketSyncDeleter) delete(ctx context.Context) (reconcile.Result, error) {
+	if bh.object.Spec.ReclaimPolicy == corev1alpha1.ReclaimDelete {
+		if err := bh.Delete(ctx); err != nil && err != storage.ErrBucketNotExist {
+			bh.object.Status.SetFailed(failedToDelete, err.Error())
+			return resultRequeue, bh.client.Status().Update(ctx, bh.object)
+		}
+	}
+	util.RemoveFinalizer(&bh.object.ObjectMeta, finalizer)
+	return reconcile.Result{}, bh.client.Update(ctx, bh.object)
+}
+
+// sync - synchronizes the state of the bucket resource with the state of the
+// bucket Kubernetes obj
+func (bh *bucketSyncDeleter) sync(ctx context.Context) (reconcile.Result, error) {
+	// crate connection secret if it doesn't exist
+	if err := bh.client.Create(ctx, bh.object.ConnectionSecret()); err != nil && !kerrors.IsAlreadyExists(err) {
+		bh.object.Status.SetFailed(failedToSaveSecret, err.Error())
+		return resultRequeue, bh.client.Status().Update(ctx, bh.object)
+	}
+
+	attrs, err := bh.Attrs(ctx)
+	if err != nil && err != storage.ErrBucketNotExist {
+		bh.object.Status.SetFailed(failedToRetrieve, err.Error())
+		return resultRequeue, bh.client.Status().Update(ctx, bh.object)
+	}
+
+	if attrs == nil {
+		return bh.create(ctx)
+	}
+
+	return bh.update(ctx, attrs)
+}
+
+// createupdater interface difning create and update operations on/for bucket resource
+type createupdater interface {
+	creater
+	updater
+}
+
+// bucketCreateUpdater implementation of createupdater interface
+type bucketCreateUpdater struct {
+	gcpstorage.Client
+	client    client.Client
+	object    *v1alpha1.Bucket
+	projectID string
+}
+
+// newBucketCreateUpdater new instance of bucketCreateUpdater
+func newBucketCreateUpdater(sc gcpstorage.Client, cc client.Client, b *v1alpha1.Bucket, pID string) *bucketCreateUpdater {
+	return &bucketCreateUpdater{
+		Client:    sc,
+		client:    cc,
+		object:    b,
+		projectID: pID,
+	}
+}
+
+// create new bucket resource and save changes back to bucket specs
+func (bh *bucketCreateUpdater) create(ctx context.Context) (reconcile.Result, error) {
+	util.AddFinalizer(&bh.object.ObjectMeta, finalizer)
+
+	if err := bh.Create(ctx, bh.projectID, v1alpha1.CopyBucketSpecAttrs(&bh.object.Spec.BucketSpecAttrs)); err != nil {
+		bh.object.Status.SetFailed(failedToCreate, err.Error())
+		return resultRequeue, bh.client.Status().Update(ctx, bh.object)
+	}
+
+	bh.object.Status.SetReady()
+
+	attrs, err := bh.Attrs(ctx)
+	if err != nil {
+		bh.object.Status.SetFailed(failedToRetrieve, err.Error())
+		return resultRequeue, bh.client.Status().Update(ctx, bh.object)
+	}
+
+	bh.object.Spec.BucketSpecAttrs = v1alpha1.NewBucketSpecAttrs(attrs)
+	if err := bh.client.Update(ctx, bh.object); err != nil {
```

Comment:
I still lean toward keeping the `client.Client` at the `Reconciler` level, limiting the scope of these CRUD functions to updating the object via its pointer. That said, using the status subresource complicates my preferred pattern when finalizers are involved because we need to figure out whether we need to update the status only, or also the metadata.

Reply:
This one is still a moving target for me. I think different controllers may use some variance in this pattern.